### DISCUSSION TITLE
5.3.0

### DISCUSSION THREAD

**Discussion by nohwnd** [pester/Pester#2052] [Releases]
_Thanks @ArmaanMcleod, @fflaten and @johlju for all your contributions and support!_

## New Code Coverage mode

New way of measuring code coverage was added. It uses the core of [Profiler](https://github.com/nohwnd/Profiler) to measure code coverage, instead of using breakpoints. This makes code coverage **much much** faster for bigger solutions. To enable it use `UseBreakpoints` option, and set it to `$false`:

```powershell
$configuration.CodeCoverage.UseBreakpoints = $false
```

The option is marked as experimental, but from our tests it looks like the new approach is as accurate as the breakpoint based approach. The new approach does not yet work on PS3.

* Add Profiler based code coverage [#1937](https://github.com/pester/pester/pull/1937)

## Configure output for failed tests

New option for controlling the verbosity of failed tests was added: 

```powershell
$configurationOutput.StackTraceVerbosity
```

With the folloing levels:

**None** - All Stack trace is hidden
**FirstLine** - First line of stack trace is shown
**Filtered** - Filtered lines that belong to Pester. This is default.
**Full** - All stack trace is shown

![None](https://user-images.githubusercontent.com/20082136/121776334-9c8bab80-cbcf-11eb-8b0a-c8eb6b1ce437.PNG)

More info and more screenshots are in the PR.
* Limit output information for failed tests [#1975](https://github.com/pester/pester/pull/1975)

## Shorter Detailed output

Output for discovery and filter was reduced in the Detailed and Diagnostic output, by moving discovery messages per file into the Diagnostic output, and hiding the Filter messages from the default Diagnostic output. All messages are still available via custom filter in `$configuration.Debug.WriteDebugMessagesFrom`.

More info and more screenshots are in the PR.
![reduced output](https://user-images.githubusercontent.com/3436158/120866791-ea316400-c590-11eb-9d68-5fde6593ed69.png)
![diagnostic output](https://user-images.githubusercontent.com/3436158/120866986-498f7400-c591-11eb-9a10-d2668252a582.png)

* Update output for Discovery and Filter [#1969](https://github.com/pester/pester/pull/1969)

## Automatic CI detection

Pester automatically detects if it is running in AzureDevOps or GitHub Actions and augments the output to report errors correctly. 

![Github actions output](https://user-images.githubusercontent.com/20082136/122943004-fe12fd80-d3b9-11eb-93a3-0206bec674b6.PNG)
![Github actions log](https://user-images.githubusercontent.com/20082136/122637969-b29bed80-d134-11eb-8096-88e8d895edc7.PNG)

The default is to autodetect CI but it can be disabled by: 

```powershell
$configuration.Output.CIFormat = 'None'
```

More info and more screenshots are in the PR.
* Azure Devops & Github Actions error logging [#1996](https://github.com/pester/pester/pull/1996)

## Override properties and methods via New-MockObject

`New-MockObject` can now override properties and methods on the created object. It counts the number of executions of a method and collects parameters provided to methods. It can also take an object as input and add properties and methods to it. 

```powershell
# Create an empty process object
$o = New-Object -TypeName 'System.Diagnostics.Process'
# Override Kill method to just return "killed"
$mockObject = New-MockObject -InputObject $o -Methods @{ Kill = { param($entireProcessTree) "killed" } }

$mockObject.Kill()
$mockObject.Kill($true)

# Inspect the history of invocations in `_<methodName>` 
$mockObject._kill

# Output:
# Call Arguments
# ---- ---------
#    1 {}
#    2 {True}
```

More examples are in the tests: https://github.com/pester/Pester/blob/main/tst/functions/New-MockObject.Tests.ps1
Real life-examples here: https://youtu.be/8GWqkGvV3H4?t=2517

* Add Ability to Override Properties on New-MockObject [#1838](https://github.com/pester/pester/pull/1838)

## Skip remaining tests when any fails

A new option was added, that allows skipping the rest of the execution when any test fails. It can skip tests in the current block and all child blocks, as well as skip all remaining tests in the file, or in the whole run.

```powershell 
$configuration.Run.SkipRemainingOnFailure = 'Block'
```

The following scopes are supported:

**Block** - Skip all remaining tests in current bloc k(including child blocks and tests) after a failed test.
**Container** - Skip all remainng tests in the container (file or scriptblock) after a failed test.
**Run** - Skip all tests across all containers in a run after a failed test.
**None** - Default, keep original behaviour.

![Run](https://user-images.githubusercontent.com/20082136/126511716-a7ee2876-8856-42b3-ae16-ee4938a9698e.PNG)

We used Skipped test result to skip those tests, but there was a lively discussion about how this should be achieved. Share your feedback if you have any.

More info and more screenshots are in the PR.
* Added Run.SkipRemainingOnFailure option [#2023](https://github.com/pester/pester/pull/2023)

## Log conflicting varibles in Mocks

Mocks pick up all variables from your tests. When your variable names are the same as parameter names of a mocked command this can lead to unexpected results in ParameterFilter. We added a new message in the Diagnostic log that warns you that a variable is in conflict with a parameter name. 

```powershell
It 'filter should not pass due to test-variable' {
    $SamAccountName = 'abc'
    # Mock parameter filter below does not pass because SamAccountName is 'def'
    Get-Stuff -SamAccountName 'def'
    # Mock parameter filter below does "incorrectly" passes because SamAccountName is not specified here
    # but is set to 'abc' by the variable above
    Get-Stuff

    # This assertion fails because there was 1 call, and not 0 calls, because we had naming conflict betwween 
    # parameters and our variables.
    Should -Invoke Get-Stuff -Times 0 -Exactly -ParameterFilter { $SamAccountName -eq 'abc' }
}
```

![log](https://user-images.githubusercontent.com/3436158/120889386-12ef4300-c5fd-11eb-83de-028abccce59d.png)

* Log conflicting variables in Should -Invoke -ParameterFilter [#1973](https://github.com/pester/pester/pull/1973)

## Better TestDrive performance

TestDrive internals were improved to save the location of test drive directory, instead of querying it every time. This greatly improved performance of any run.

When you don't need `TestDrive` or `TestRegistry`, you can disable both TestDrive and TestRegistry entirely, to get more performance gains: 

```powershell
$configuration.TestDrive.Enabled = $false
$configuration.TestRegistry.Enabled = $false
```

Disabling TestRegistry is also useful on restricted systems when access to Windows Registry is disabled by group policy. 

* Use profiler to improve perf of test drive, name expanding and add-member [#1955](https://github.com/pester/pester/pull/1955)
* Add option to disable TestDrive and TestRegistry [#2008](https://github.com/pester/pester/pull/2008)

## Additional fixes in documentation, output, and other bugs squashed:

* Fixed formatting for null elements in an array [#1961](https://github.com/pester/pester/pull/1961)
* Make InModuleScope-parameters available as variables in scriptblock [#1957](https://github.com/pester/pester/pull/1957)
* Improve operator-help returned from Get-ShouldOperator [#1988](https://github.com/pester/pester/pull/1988)
* Added Pester version message to detailed and diagnostic output [#1962](https://github.com/pester/pester/pull/1962)
* Removed StartMessage from output [#1968](https://github.com/pester/pester/pull/1968)
* Add formatdata to ensure list view for all sections in PesterConfiguration [#1964](https://github.com/pester/pester/pull/1964)
* Update reporttheme output to remove hard coded colour names [#1934](https://github.com/pester/pester/pull/1934)
* Set name in xml element testsuite [#2034](https://github.com/pester/pester/pull/2034)
* Added FileContentMatchMultilineExactly should assertion [#2030](https://github.com/pester/pester/pull/2030)
* Added Filter.ExcludeLine option [#2016](https://github.com/pester/pester/pull/2016)
* Add PSObject boxing support in PesterConfiguration [#1978](https://github.com/pester/pester/pull/1978)
* Add missing container-parameters to Data [#1986](https://github.com/pester/pester/pull/1986)
* Update InvokeVerifiable error message and add negate support [#1976](https://github.com/pester/pester/pull/1976)
* Adds support for Register-ArgumentCompleter [#1979](https://github.com/pester/pester/pull/1979)
* Fix throwing on missing folders [#2012](https://github.com/pester/pester/pull/2012)
* make Value-property setter private in options [#1994](https://github.com/pester/pester/pull/1994)
* Add analyzerrule for object-commands [#1753](https://github.com/pester/pester/pull/1753)
* Update New-PesterConfiguration help [#1946](https://github.com/pester/pester/pull/1946)
* Fix session state in Should -Invoke to use the correct session state [#1954](https://github.com/pester/pester/pull/1954)
* Update Pester.RSpec.ps1 [#1951](https://github.com/pester/pester/pull/1951)
* Save all files with UTF8BOM [#1950](https://github.com/pester/pester/pull/1950)
* Fix formatting in all files [#1949](https://github.com/pester/pester/pull/1949)
* Allow positional parameter in get-shouldoperator [#1947](https://github.com/pester/pester/pull/1947)
* Remove about links they break website MDX [#1948](https://github.com/pester/pester/pull/1948)
* Added missing format tests [#1935](https://github.com/pester/pester/pull/1935)

See full log [here](https://github.com/pester/pester/compare/5.2.2...5.3.0)

<hr /><em>This discussion was created from the release <a href='https://github.com/pester/Pester/releases/tag/5.3.0'>5.3.0</a>.</em>

**Comment by fred-gagnon** [comment]
Indeed, the new Profiler is much faster than the breakpoint method. Our unit tests now run in about 10 minutes versus 17 minutes. However, I am seeing a few differences in coverage between the two methods. After processing the coverage file with Report Generator, I observe the following discrepency (on the left is the coverage using the Profiler while on the right is the coverage using Breakpoints):

![image](https://user-images.githubusercontent.com/61014947/131678129-aaff3b3a-0ad4-440b-bf03-27e0737b1aa4.png)

What the profiler is reporting for these lines doesn't make much sense.

**Reply by nohwnd** [reply]
it's probably some special case I am not taking into account where the reported hit needs to be translated to a different number to agree with what the AST says is a command that should run. I will put this into tests and see what's wrong :) 

Or you can try it: This is the source file that has edge cases I found so far for code coverage: 

https://github.com/pester/Pester/blob/main/tst/CoverageTestFile.ps1

And these are tests that compare if the coverage is the same as before, there are also some negative cases where the previous coverage is less accurate, and those are commented out in the CoverageTestFile.ps1 source file: 

https://github.com/pester/Pester/blob/main/tst/Pester.RSpec.Coverage.ts.ps1

**Reply by nohwnd** [reply]
If you'd decide to PR this, please target the `rel/3.1.x` branch 🙂
